### ISSUE TITLE
Would adding support for RDS instances be trivial? 

### ISSUE THREAD

**Issue by daveseff** [puppetlabs/puppetlabs-aws#41]
Would like to be able to manage RDS instances. Any plans for adding provider/rds_instance support?

**Comment by garethr** [comment]
Hi @daveseff. We'd definitely like to support RDS, although we have a few other resources we're working on first. Would you be able to describe which parts of RDS you're trying to use, and help us identify a minimum property set for the type?

**Comment by daveseff** [comment]
I'm looking to spin up MySQL instances, So I believe I could do something like :

rds_instance {'somedumbdb' :
   db_instance_id => 'My Database',
   db_user => ''admin',
   db_pass => 'somepassword',
   db_engine => 'mysql', 
   db_version => '5.6.7', 
   db_class => 'db.t2.micro', 
   db_multi_az => 'yes', 
   db_storage => 30,   # in GB
}

**Comment by garethr** [comment]
Thanks, that's super helpful. I can't say exactly when we'll get to this, but if it's just the single resource that makes it simpler. Does this need a security group as well?

**Comment by daveseff** [comment]
Thanks. I believe it does. 

I'm rrying to get more familiar with the code also to help extend the module's ability. My ruby is rusty. I'm more python :)

**Comment by garethr** [comment]
@daveseff if we can help let me know. Happy to provide pointers or just on IRC to answer questions or look over not quite working code.

**Comment by daveseff** [comment]
I have a fork in progress. I've used ec2_instance provider as a starting point. I think I have a hande on most of it, I just dont know what I need to do to this section to adapt for RDS. Ahy pointers? 

def self.instances
    regions.collect do |region|
      instances = []
      rds_client(region).describe_db_instances(filters: [
        {name: 'MaxRecords', values: ['100', ]}
      ]).each do |response|
        response.data.reservations.each do |reservation|
          reservation.instances.each do |instance|
            hash = instance_to_hash(region, instance)
            instances << new(hash) if (hash[:name] and ! hash[:name].empty?)
          end
        end
      end
      instances
    end.flatten
  end

**Comment by garethr** [comment]
If you can share your fork I'll probably be able to provide better pointers, but the code should probably look like:

```ruby
def self.instances
    regions.collect do |region|
      instances = []
      rds_client(region).describe_db_instances.each do |response|
        response.data.db_instances.each do |db|
            hash = db_instance_to_hash(region, db)
            instances << new(hash) if hash[:name]
          end
        end
      end
      instances
    end.flatten
  end
```

Note that calling `each` on `describe_db_instances` will automatically loop over all instances, so you probably don't need the filter.

The main work will be done in the `db_instance_to_hash` method.

According to the docs, the filters keyword isn't currently supported either: http://docs.aws.amazon.com/sdkforruby/api/frames.html

**Comment by daveseff** [comment]
Thanks. My working fork is here:

https://github.com/daveseff/puppetlabs-aws

It seems to create the DBs now. Although I see an error about 'master_user_password'. No I'm needing to fix that error and add more functionality. 

The puppet test.pp file I've been testing with is 
`
RDS_instance {
   region            => 'us-west-1',
   availability_zone_name => 'us-west-1a',
}

rds_instance { 'testdb':
  region => 'us-west-1',
  db_name => 'testdb',
  engine => 'mysql',
  engine_version => '5.6.19a',
  license_model => 'general-public-license',
  allocated_storage => 10,
  availability_zone_name => 'us-west-1a',
  storage_type => 'gp2',
#  iops => 1000, 
  db_instance_class => 'db.t2.micro',
  master_username => 'admin',
  master_user_password => 'thepassword',
  multi_az => false,
  db_subnet_group_name => 'tools-db',
#  security_groups => 'tools-db', 
}
`

**Comment by abuxton** [comment]
@daveseff  I spoke to Gareth and he said you were coming along with this functionality.
Just wondering how this is progressing, I'm with a group looking for this functionality, be interested to find out how its working for you.
Also have you had any issue with the time to spin up RDS instances.

**Comment by daveseff** [comment]
@abuxton I do have a fork over at https://github.com/daveseff/puppetlabs-aws. I can create and destroy RDS instances just fine. I am, however having an issue with having this in my manifests because I am having difficulty getting puppet to recognize that the RDS instance already has been provisioned and complains. Now that I'm back from Shmoocon I can take another look. Any hints here are welcome :)

Error: db_instance_class property is read-only once rds_instance created.
Error: /Stage[main]/Etouches_tools/Rds_instance[test-jira]/db_instance_class: change from absent to db.t2.micro failed: db_instance_class property is read-only once rds_instance created.

**Comment by daveseff** [comment]
@abuxton   Still learning how this works as I go. I've never written anything this complex in puppet yet. :) 

I fixed the above issue. Still trying to figure things out.

**Comment by petems** [comment]
@daveseff Hello! Thanks for the PR, this is exactly what me and @abuxton were looking for! :smile: 

So the issue you were running into was that db_name is deceptively named, it's actually the name of the default database on the server when it's created. The ID used to reference the DB is actually the `db_instance_identifier` field.

I've taken your code and figured some stuff out, implemented the commented out tag codes and I'm planning on writing tests over the weekend.

Check it out here: https://github.com/petems/puppetlabs-aws/tree/add_RDS_support

I've also created a modules ticket to keep track of stuff here: https://tickets.puppetlabs.com/browse/MODULES-1702

When I've finished, do you mind if I open a new PR with your commits in it? Or do you want me to PR into your repo and we can merge from the original PR? :+1:

**Comment by daveseff** [comment]
@petems  Thanks. I don't mind if you open a new PR with your fixes. 
On a similar topic I've been adding support to create ec2_instances for VPCs. It was an easy fix and will create a PR for that once the RDS stuff is all taken care of. .

**Comment by petems** [comment]
@daveseff That would be awesome, for now I've been testing against `m3` because it was the smallest server you can make without a subnet...

**Comment by garethr** [comment]
Hi @daveseff, you can see work on supporting attaching instances to subnets in https://github.com/puppetlabs/puppetlabs-aws/pull/61. Their are some subtleties to that PR that might interest, also it's a good example of the tests we're adding for new features.

**Comment by danieldreier** [comment]
Is there any way we can mange feature groups, as well? Based on my experiments with RDS, Drupal performance under stock RDS mysql settings is unsatisfactory, but setting half a dozen settings in a parameter group makes it totally usable. I'm just imagining passing a key => value hash to an rds_property_group resource and then assigning the rds instance to a property group as a property.

(I'm willing to work on a PR if needed but thought I'd put my feature request in here in case somebody else gets to it first; it's a little awkward to do PRs against PRs.

**Comment by petems** [comment]
@danieldreier I did some similar stuff for `rds_db_securitygroup` as a type and provider in the instance PR, but that was needed as an MVP to use Puppet to create an RDS instance that's actually reachable: https://github.com/petems/puppetlabs-aws/commit/768894a11492f192c5f961365f7441dc304022a7

I think the best approach would be to add an `rds_parameter_group` as a resource in a separate PR, as it doesn't actually require the instance work to create the type and provider :+1: If you do a PR I wouldn't mind having a look and trying to the other branches code against it to test it out! :+1:

**Comment by petems** [comment]
I pulled the code into a seperate an RDS type and provider: https://github.com/petems/petems-rds, will deprecate when merged :smile:

**Comment by activars** [comment]
I wonder if this would support provision RDS users & permission?

**Comment by petems** [comment]
@daveseff Your code (with some modifications by myself and @garethr) has been merged! Keep an eye out for a release soon-ish! :eyes:
